react/features/video-layout/subscriber.js

```diff
@@ -1,24 +1,105 @@
 // @flow
-
 import {
     VIDEO_QUALITY_LEVELS,
     setMaxReceiverVideoQuality
 } from '../base/conference';
-import { StateListenerRegistry } from '../base/redux';
+import {
+    getPinnedParticipant,
+    pinParticipant
+} from '../base/participants';
+import { StateListenerRegistry, equals } from '../base/redux';
 import { selectParticipant } from '../large-video';
 import { shouldDisplayTileView } from './functions';
+import { setParticipantsWithScreenShare } from './actions';
+
+declare var interfaceConfig: Object;
 
 /**
  * StateListenerRegistry provides a reliable way of detecting changes to
  * preferred layout state and dispatching additional actions.
  */
 StateListenerRegistry.register(
     /* selector */ state => shouldDisplayTileView(state),
-    /* listener */ (displayTileView, { dispatch }) => {
+    /* listener */ (displayTileView, store) => {
+        const { dispatch } = store;
+
         dispatch(selectParticipant());
 
         if (!displayTileView) {
-            dispatch(setMaxReceiverVideoQuality(VIDEO_QUALITY_LEVELS.HIGH));
+            dispatch(
+                setMaxReceiverVideoQuality(VIDEO_QUALITY_LEVELS.HIGH));
+
+            _updateAutoPinnedParticipant(store);
+        }
+    }
+);
+
+/**
+ * For auto-pin mode, listen for changes to the known media tracks and look
+ * for updates to screen shares.
+ */
+StateListenerRegistry.register(
+    /* selector */ state => state['features/base/tracks'],
+    /* listener */ (tracks, store) => {
+        if (typeof interfaceConfig !== 'object'
+            && !interfaceConfig.AUTO_PIN_LATEST_SCREEN_SHARE) {
+            return;
+        }
+
+        const oldSharingParticipantIds
+            = store.getState()['features/video-layout'].screenShares || [];
+        let newSharingParticipantIds = Array.from(oldSharingParticipantIds);
+        const currentDesktopStreams = tracks.filter(track =>
+            track.mediaType === 'video' && track.videoType === 'desktop');
+
+        // Make sure all new desktop streams get added to the known screen
+        // shares.
+        currentDesktopStreams.forEach(({ participantId }) => {
+            if (!newSharingParticipantIds.includes(participantId)) {
+                newSharingParticipantIds.push(participantId);
+            }
+        });
+
+        // Filter out any participants which are no longer screen sharing
+        // by looping through the desktop streams and making sure only those
+        // participant IDs with a desktop stream are included.
+        newSharingParticipantIds = newSharingParticipantIds.filter(participantId =>
+            currentDesktopStreams.find(stream => stream.participantId === participantId));
```

Comment:
If you use `.map` to get an array with the `participantId`s from the `currentDesktopStreams`, isn't that going to be the `newSharingParticipantIds` array? Or this will break the order?

Even if it breaks the order I guess having the array of `participantId`s can be used to replace the `.find` with `includes`. WDYT?

Reply:
I'm now thinking to use Array.prototype.reduce to get the participant ids from the tracks store.